src/qutip_qip/circuit/circuitsimulator.py

```diff
@@ -599,41 +498,61 @@ def _evolve_state(self, U):
         U: Qobj
             unitary to be applied.
         """
-
+        if operation.name == "GLOBALPHASE":
+            # This is just a complex number.
+            U = np.exp(1.0j * operation.arg_value)
+        else:
+            # We need to use the circuit because the custom gates
+            # are still saved in circuit instance.
+            # This should be changed once that is deprecated.
+            U = self.qc._get_gate_unitary(operation)
+            U = expand_operator(
+                U,
+                dims=self.dims,
+                targets=operation.get_all_qubits(),
+            )
         if self.mode == "state_vector_simulator":
-            self._evolve_ket(U)
+            state = U * state
         elif self.mode == "density_matrix_simulator":
-            self._evolve_dm(U)
+            state = U * state * U.dag()
         else:
             raise NotImplementedError(
                 "mode {} is not available.".format(self.mode)
             )
+        return state
+
+    def _evolve_state_einsum(self, gate, state):
+        if gate.name == "GLOBALPHASE":
+            # This is just a complex number.
+            return np.exp(1.0j * gate.arg_value) * state
+        # Prepare the state tensor.
+        targets_indices = gate.get_all_qubits()
+        if isinstance(state, Qobj):
+            # If it is a Qobj, transform it to the array representation.
+            state = state.full()
+            # Transform the gate and state array to the corresponding
+            # tensor form.
+            state = state.reshape(self._tensor_dims)
+        # Prepare the gate tensor.
+        gate = self.qc._get_gate_unitary(gate)
+        gate_array = gate.full().reshape(gate.dims[0] + gate.dims[1])
+        # Compute the tensor indices and call einsum.
+        num_site = len(state.shape)
+        ancillary_indices = range(num_site, num_site + len(targets_indices))
+        index_list = range(num_site)
+        new_index_list = list(index_list).copy()
```

Comment:
```suggestion
        new_index_list = list(index_list)
```
Since `index_list` is a range, `list(index_list)` is already a new list.